### DISCUSSION TITLE
At the risk of someone laughing?  Anyone implement a speed test in MP?

### DISCUSSION THREAD

**Discussion by Jibun-no-Kage** [micropython/micropython#9179] [General]
At the risk of someone laughing?  Any one implement a speed test in MP?  I know, this maybe insane, but I have a need or use case, to validate one of my routers which seems suspect, and as I was setting up a Pi with the speedtest-cli and getting ready to do some stress testing... the (crazy) thought came to mind, it would be kinda cool to do this on a Esp32 or Pico, no?  So, hence the question asked.  Maybe something as complex as speedtest-cli is not possible or realistic on a microcontroller, but maybe a simpler test concept?

**Comment by stinos** [comment]
Problem for microcontrollers is your speed might very well be limited by the device's hardware itself, not the network. On the other hand if you want to know that exactly i.e. how fast you can send/receive on the device then it could be useful. But then it still depends on the exact implementation what the numbers mean. I can imagine writing up something in MicroPython might not be as performant as in plain C.

**Comment by dpgeorge** [comment]
If you can control both ends (eg a server somewhere on the internet) then you can use `iperf3` to test TCP and UDP bandwidth: https://github.com/micropython/micropython-lib/tree/master/python-ecosys/iperf3

**Comment by Jibun-no-Kage** [comment]
@stinos, this is not to get a true speed, but to set a qualified baseline.  Whatever it is, and then over time, watch for a significant delta.  The goal is consistency more than actual performance.

**Comment by Jibun-no-Kage** [comment]
@dpgeorge, actually given my network setup I likely could use a internal target.  Thinking about it.  We did something like this at the enterprise design lab where I worked before I retired.  We had a specific location in our intranet, that we knew the context of its location, performance, etc. controlled resource loading.  So we often directed a benchmark testing to that specific site.  True we had some variance across the intranet, but because we know the day of week, day of month, 24 hour loading context, we knew with reasonable expectation when we could run stress and performance tests, with the least number of random factors.

**Reply by scruss** [reply]
Using _iperf_ is described on page 25 of the [Connecting to the Internet with Raspberry Pi Pico W](https://datasheets.raspberrypi.com/picow/connecting-to-the-internet-with-pico-w.pdf) document

**Comment by karfas** [comment]
Not counting the coolness/fun/exercise factor, I don't see the value in this project.

One of the "large" RaspPIs is much better equipped for doing this kind of tasks, is also not expensive and also can run micropython.

**Comment by Jibun-no-Kage** [comment]
Yup, that is why I stated the topic the way I did.  This effort has little practical benefit in a general context, I mean either your WiFi is stable or not, either you get a given performance that is reasonable or not (given all the factors that impact WiFi performance).  The only possible benefit is that for a given device, at a given location (in your house, home, work site, etc.) you know said device works at X through-put or consistency, and how it trends over time, or every time you invoke said test.  

Using my cell phone, initially locating the best locations for my routers, i.e. zones, worked, there a number of ways and apps to help with this.  You get some odd results depending on the construction of the building in question.  For example, in my garage, the nearest router is not usually the best connection overall, oddly the router in my casitas just happens to be better... for my cell phone.  Even though it is at a longer distance to the garage, but there is/are less physical structure that absorbs signal/strength.  Whereas the Pi device that controls stuff in my garage, and has a few sensors, that monitor things in my garage, seems to like the router in the main house better than the casitas.  Go figure.

Given that micro-controllers have a highly variable range of WiFi behavior, being able to use them as a test node in of its self has merit.  Using a cell phone or a Pi device may do better, but that is not the goal, the goal is to use the actual device, i.e. the specific micro-controller, in the actual context in which it maybe placed or used.  Such as in my garage, will the micro-controller do better with my casitas router or the main house router?  This project could provide insight into that question, at least to my specific use case.
